zapcore/write_syncer.go

```diff
@@ -75,6 +78,87 @@ func (s *lockedWriteSyncer) Sync() error {
 	return err
 }
 
+type bufferWriterSyncer struct {
+	ws           WriteSyncer
+	bufferWriter *bufio.Writer
+}
+
+// defaultBufferSize sizes the buffer associated with each WriterSync.
+const defaultBufferSize = 256 * 1024
+
+// defaultFlushInterval means the default flush interval
+const defaultFlushInterval = 30 * time.Second
+
+// CancelFunc should be called when the caller exits to clean up buffers.
+type CancelFunc func() error
+
+// Buffer wraps a WriteSyncer in a buffer to improve performance,
+// if bufferSize = 0, we set it to defaultBufferSize
+// if flushInterval = 0, we set it to defaultFlushInterval
+func Buffer(ws WriteSyncer, bufferSize int, flushInterval time.Duration) (WriteSyncer, CancelFunc) {
+	ctx, cancel := context.WithCancel(context.Background())
+
+	cancelfunc := func() error {
+		cancel()
+		return ws.Sync()
```

Comment:
we should collect errors that may have caused the background goroutine to have ended already too

Reply:
we do not need the make things so complicated, just simply get errors from Sync() func instead.

and the background goroutine just keep syncing until the cancel func is called.

Reply:
Calling `Sync` and returning that error makes sense, thanks.